stackslib/src/chainstate/burn/sortition.rs

```diff
@@ -269,6 +283,219 @@ impl BlockSnapshot {
+        let Some(one_minus_atc) = AtcRational::one().sub(&atc) else {
+            // somehow, ATC > 1.0, then miners spent more than they did in the last sortition.
+            // So, the null miner loses.
+            warn!("ATC > 1.0 ({})", &atc.to_hex());
+            return AtcRational::zero();
+        };
+
+        let Some(atc_prod_adv) = atc.mul(&adv) else {
+            // if this is somehow too big (impossible), it would otherwise imply that the null
+            // miner advantage is overwhelming
+            warn!("ATC * ADV == INF ({} * {})", &atc.to_hex(), &adv.to_hex());
+            return AtcRational::one();
+        };
+
+        let Some(sum) = one_minus_atc.add(&atc_prod_adv) else {
+            // if this is somehow too big (impossible), it would otherwise imply that the null
+            // miner advantage is overwhelming
+            warn!(
+                "(1.0 - ATC) + (ATC * ADV) == INF ({} * {})",
+                &one_minus_atc.to_hex(),
+                &atc_prod_adv.to_hex()
+            );
+            return AtcRational::one();
+        };
+        sum.min(&AtcRational::one())
+    }
+
+    /// Determine whether or not the null miner has won sortition.
+    /// This works by creating a second burn distribution: one with the winning block-commit, and
+    /// one with the null miner.  The null miner's mining power will be computed as a function of
+    /// their ATC advantage.
+    fn null_miner_wins(
+        sort_tx: &mut SortitionHandleTx,
+        block_header: &BurnchainBlockHeader,
+        sortition_hash: &SortitionHash,
+        commit_winner: &LeaderBlockCommitOp,
+        atc: AtcRational,
+    ) -> Result<bool, db_error> {
+        let vrf_seed = Self::get_last_vrf_seed(sort_tx, block_header)?;
+
+        let mut null_winner = commit_winner.clone();
+        null_winner.block_header_hash = {
+            // make the block header hash different, to render it different from the winner.
+            // Just flip the block header bits.
+            let mut bhh_bytes = null_winner.block_header_hash.0.clone();
+            for byte in bhh_bytes.iter_mut() {
+                *byte = !*byte;
+            }
+            BlockHeaderHash(bhh_bytes)
+        };
+
+        let mut null_sample_winner = BurnSamplePoint::zero(null_winner.clone());
+        let mut burn_sample_winner = BurnSamplePoint::zero(commit_winner.clone());
+
+        let null_prob = Self::null_miner_probability(atc);
+
+        // AtcRational's integer part is only 64 bits, so we need to scale it up so that it occupes the
+        // upper 64 bits of the burn sample point ranges so as to accurately represent the fraction
+        // of mining power the null miner has.
+        let null_prob_u256 = if null_prob.inner() >= AtcRational::one().inner() {
+            // prevent left-shift overflow
+            AtcRational::one_sup().into_inner() << 192
+        } else {
+            null_prob.into_inner() << 192
+        };
```

Comment:
This is still a leaky abstraction (making the inner 256 bits of AtcRational private doesn't matter if the consumer still has to reason about what exactly AtcRational is doing with its bits). It should be pretty trivial to avoid doing this (and make this code more clear). This comment suggests that what you care about here is the integer part of the AtcRational, but that doesn't seem to be the case: you actually want the fractional part of the AtcRational to occupy the whole 256 bit range used for the burn sample (which is what you're doing by left shifting 192 bits -- the high 192 bits of AtcRational are the overflow bits and the integer parts, unless I'm really misreading things).

Suggestion:

```suggestion
        // Compute null probability in 256 bit space 
        let null_prob_u256 = if null_prob >= AtcRational::one() {
            // is there a test for this case?
            Uint256::max()
        } else {
            Uint256::from_u64(null_prob.fpart()) << 192
        };
```

This way, the reader doesn't need to know both what the inner byte representation of AtcRational is and how the 256 bit math impacts the burn sample points.

Reply:
I'm trying to avoid a lot of essentially needless boilerplate trait impls for operator overloading.  But I can add a method to construct a "clamped" `Uint256` from an `AtcRational` (which is what's really happening here).

Reply:
Okay, I hid this behind a method `AtcRational::into_sortition_probability()`